src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -2100,46 +2625,79 @@ UCS_PROFILE_FUNC_ALWAYS(ucs_status_t, uct_ib_mlx5_devx_reg_exported_key,
                         (md, memh), uct_ib_mlx5_md_t *md,
                         uct_ib_mlx5_devx_mem_t *memh)
 {
-    struct mlx5dv_devx_obj *cross_mr;
+    struct mlx5dv_devx_obj *cross_mr = NULL;
+    uct_ib_mlx5_devx_umr_t *umr      = NULL;
+    uct_ib_mlx5_devx_mr_t *mr        = &memh->mrs[UCT_IB_MR_DEFAULT];
     uint32_t exported_lkey;
     ucs_status_t status;
 
     if (memh->dm != NULL) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    ucs_assertv(memh->cross_mr == NULL,
-                "memh=%p cross_mr=%p exported_lkey=0x%x", memh, memh->cross_mr,
-                memh->exported_lkey);
+    ucs_assertv((memh->umr_mr == NULL) && (memh->cross_mr == NULL),
+                "memh=%p umr_mr=%p cross_mr=%p exported_lkey=0x%x", memh,
+                memh->umr_mr, memh->cross_mr, memh->exported_lkey);
     if (!(md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
         goto out_umem_mr;
     }
 
-    ucs_assert(!(memh->super.flags & UCT_IB_MEM_MULTITHREADED));
-    status = uct_ib_mlx5_devx_reg_ksm_data_contig(md, memh->address,
-                                                  (uint64_t)memh->address, 0, 0,
-                                                  "exported-key",
-                                                  &memh->mrs[UCT_IB_MR_DEFAULT],
-                                                  &cross_mr, &exported_lkey);
-    if (status != UCS_OK) {
-        return status;
+    /* Current bind impl (IBV_WR_BIND_MW) attaches a single KLM segment, so:
+     * - IBV_WR_BIND_MW supports the maximum region length of 2GB
+     * - IBV_WR_BIND_MW does not support multi-segment (multi-threaded) MRs
+     * For these use cases we fallback to KSM
+     */
+    if ((mr->super.ib->length > UCT_IB_MD_MAX_MR_SIZE) ||
+        (memh->super.flags & UCT_IB_MEM_MULTITHREADED)) {
+        status = uct_ib_mlx5_devx_reg_ksm_data(md, memh, UCT_IB_MR_DEFAULT, 0,
+                                               0, 0, "exported-key", &cross_mr,
+                                               &exported_lkey);
+        if (status != UCS_OK) {
+            return status;
+        }
+    } else {
+        status = uct_ib_mlx5_devx_umr_mkey_bind(md, mr, &umr);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        exported_lkey = umr->umr_mkey->lkey;
+
+        /* Return if mkey is already marked as XGVMI */
+        if (uct_ib_mlx5_devx_umr_mkey_is_xgvmi(exported_lkey)) {
```

Comment:
it which case it's not XGVMI?

Reply:
It's not XGVMI allowed when we are creating a new UMR mkey, not reusing existing mkey from the pool. For such a newly created key we need to explicitly allow XGVMI with `uct_ib_mlx5_devx_allow_xgvmi_access` call below.

Reply:
Ok, there are 2 ways to check it directly:
1. Out param `int *from_pool`, in the `uct_ib_mlx5_devx_umr_mkey_bind` function, indicating that returned mkey is from the pool. This was the initial impl
2. Add field `int xgvmi_allowed` to the `uct_ib_mlx5_devx_umr_mkey_t`. This is the most explicit way, we discussed it with Mikhail. The only downside is that it consumes a bit more memory comparing to current implicit impl. But ok

I guess you prefer the 2nd option, so I'm implementing it

Reply:
I refactored existing code to split the flows:
- Separate flow for KSM exported key
- Separate flow for UMR exported key - now we create valid mkey and allow xgvmi during creation.
- KSM or UMR may reset XGVMI capability flag if fail to allow xgvmi -> in this case we fallback to UMEM regsitration